### ISSUE TITLE
[Feature Request] Preview auto scrolls to same location your cursor is at in the editor

### ISSUE THREAD

**Issue by moorepants** [retext-project/retext#222]
With a long rst document I have to constantly scroll the preview pane to the area where I'm editing. It would be nice if the preview would track where you are editing and go to that spot on reload to prevent the user from having to manually scroll the preview window.

**Comment by mitya57** [comment]
We currently have synchronized scrolling for Markdown (since ReText 6.0), but it's not trivial to implement this for reStructuredText…

**Comment by crystalfp** [comment]
Sorry, but with 6.0.2 on Windows the live preview is not synchronized.
The  "Enable synchronized scrolling for markdown" is checked, the Default markup is set to Markdown and my file has extension .md and is UTF-8 Markup.
Is there anything other that I have to check?
Thanks!
mario

**Comment by mitya57** [comment]
@crystalfp 
> Is there anything other that I have to check?

Yes. Is the WebKit renderer enabled? (Note that the official Qt installers no longer include a WebKit module, so you may need to build it manually…)

**Comment by crystalfp** [comment]
No, it is not enabled. It is grayed out. 
What should I do to build it? 
Thanks! 
Mario

Il 30 Novembre 2016 15:08:32 CET, Dmitry Shachnev <notifications@github.com> ha scritto:
>@crystalfp 
>> Is there anything other that I have to check?
>
>Yes. Is the WebKit renderer enabled? (Note that the official Qt
>installers no longer include a WebKit module, so you may need to build
>it manually…)
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/retext-project/retext/issues/222#issuecomment-263881827

-- 
Inviato da Galaxy Notes con K-9 Mail. Perdonate la brevità.

**Comment by mitya57** [comment]
> What should I do to build it?

I do not use Windows, so I cannot give you the precise instructions. But you can probably follow instructions on [this wiki page](https://github.com/annulen/webkit/wiki/Building-QtWebKit-on-Windows) (and source code from [the same repository](https://github.com/annulen/webkit)). Note that in will take a lot of time (> 1 hour), lots of RAM (> 4 GB) and also some disk space (~ 10 GB).

We are investigating switching to Qt WebEngine (which is shipped with Qt), but that won't happen soon.

**Comment by crystalfp** [comment]
Sigh, too much work.
So I return to my trusted Markdown workflow: Notepad++ for writing and WriteMonkey for writing &  rendering.
Anyway, thanks!
mario

**Comment by mitya57** [comment]
> Sigh, too much work.

You can try to use the pre-compiled (MSVC 2015) qtwebkit binaries from https://github.com/annulen/webkit/releases/tag/qtwebkit-tp4. You will still need to compile the Python bindings yourself (but this is much easier to do).

Anyway, ReText is in no way officially supported on Windows, and on popular Linux distros both qtwebkit and its bindings are available from the repositories.

**Comment by thomashoddinott** [comment]
> We currently have synchronized scrolling for Markdown (since ReText 6.0), but it's not trivial to implement this for reStructuredText…

Any chance we're going to see a `Enable synchronised scrolling for reStructuredText` option anytime?

**Comment by mitya57** [comment]
Hi @thomashoddinott!

> Is there any chance we're going to see a Enable synchronised scrolling for reStructuredText in the future?

Unfortunately, I didn't have enough time for that yet. And there is also a technical problem: synchronized scrolling requires that the generated HTML contains line numbers information, and not all nodes provide this information: https://stackoverflow.com/q/58359284.

> I'm working with a lot of math blocks. With use webkit renderer enabled I'm able to see beautiful rendered previews and double check my equations, but as soon as I make an edit, my preview screen jumps back to the top, making ReText difficult to use.

This problem can be fixed without implementing proper *synchronized* scrolling.

I have just pushed commit a2b4beffb4311ad146ee81de4b2e68839ee3265b which should make it better. For me, it fixes jumping to top when I edit reStructuredText documents with math. Can you please test it?

**Comment by thomashoddinott** [comment]
Hi @mitya57,

I tested it today and it's working.

Thanks for your help.

**Comment by mitya57** [comment]
With the latest commit in pymarkups module, synchronized scrolling should be working with reStructuredText too (provided that WebKit or WebEngine renderer is used).

You can test it with the following command:

    pip3 install --user --upgrade https://github.com/retext-project/pymarkups/archive/master.tar.gz

Any feedback is welcome!

**Comment by gbastien** [comment]
Hi,

just to say that I was looking why synch scrolling was not working with reStructuredText, I installed pymarkups as suggested here above and now, when enabling "webkit rendering" the auto scroll is working as expected.

Great tool!

Gauthier
